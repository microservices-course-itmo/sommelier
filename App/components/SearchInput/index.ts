export { default as SearchInput, SearchInputProps } from './input'
