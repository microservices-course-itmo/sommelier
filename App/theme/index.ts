import { headerSizes } from './headerSizes'
import { boxSizes } from './boxSizes'
import { colors } from './colors'
import { fontSizes } from './fontSizes'
import { borders } from './borders'
import { inputSizes, searchSizes, inputFontSizes } from './inputSearchSizes'

const theme = {
  boxSizes: {
    ...boxSizes,
  },
  headerSizes: {
    ...headerSizes,
  },
  colors: {
    ...colors,
  },
  fontSizes: {
    ...fontSizes,
  },
  borders: {
    ...borders,
  },
  inputSizes: {
    ...inputSizes,
  },
  searchSizes: {
    ...searchSizes,
  },
  inputFontSizes: {
    ...inputFontSizes,
  },
}

export default theme
