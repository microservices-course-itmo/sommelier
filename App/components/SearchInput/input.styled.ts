import { TextInputProps } from 'react-native'
import styled from 'styled-components/native'
import theme from '../../theme/index'

type Props = {
  size: keyof typeof theme.inputSizes
}

type ViewProps = {
  editable: boolean
}

export const StyledInput = styled.TextInput<Props & TextInputProps>`
  font-size: ${({ size, theme }) => theme.inputSearchFontSizes[size]};
  padding-left: 5px;
  padding-right: 60px;
  width: 100%;
  height: ${({ size, theme }) => theme.inputSearchSizes[size]};
`

export const StyledView = styled.View<ViewProps>`
  display: flex;
  flex-direction: row;
  border: 1px solid ${theme.colors.searchInputBorder};
  border-radius: 50px;
  align-items: center;
  overflow: hidden;
  padding-left: 15px;
  padding-right: 5px;
  padding-top: 2px;
  padding-bottom: 2px;
  width: 100%;
  background-color: ${({ editable }) =>
    editable ? theme.colors.white : theme.colors.notEditableField};
`

export const ViewClose = styled.View`
  position: absolute;
  right: 15px;
  height: 100%;
  justify-content: center;
  margin-left: 5px;
`
