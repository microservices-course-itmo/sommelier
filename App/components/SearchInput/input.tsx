import React, { useState } from 'react'
import {
  NativeSyntheticEvent,
  TextInputChangeEventData,
  TouchableOpacity,
} from 'react-native'
import theme from '../../theme/index'
import * as Styled from './input.styled'
import Search from '../../../assets/images/search.svg'
import Close from '../../../assets/images/close.svg'
import { INPUT_SIZE_KEYS } from '../../theme/inputSearchSizes'

type Props = {
  defaultValue?: string
  placeholder?: string
  editable?: boolean
  size?: keyof typeof theme.inputSearchSizes
  maxLength?: number
  value?: string
  onChange: (e: NativeSyntheticEvent<TextInputChangeEventData>) => void
  clearButtonMode?: boolean
}

const SearchInput = ({
  defaultValue,
  placeholder,
  editable = true,
  size = INPUT_SIZE_KEYS.md,
  value,
  onChange,
  clearButtonMode,
  maxLength,
}: Props) => {
  const [text, setText] = useState(value)
  return (
    <Styled.StyledView editable={editable}>
      <Search
        width={theme.searchSizes[size]}
        height={theme.searchSizes[size]}
      />
      <Styled.StyledInput
        defaultValue={defaultValue}
        placeholder={placeholder}
        editable={editable}
        size={size}
        value={text}
        onChange={onChange}
        onChangeText={(text) => setText(text)}
        maxLength={maxLength}
      />
      {clearButtonMode && (
        <Styled.ViewClose>
          <TouchableOpacity onPress={() => setText('')}>
            <Close
              width={theme.inputSearchFontSizes[size]}
              height={theme.inputSearchFontSizes[size]}
            />
          </TouchableOpacity>
        </Styled.ViewClose>
      )}
    </Styled.StyledView>
  )
}

export default SearchInput
