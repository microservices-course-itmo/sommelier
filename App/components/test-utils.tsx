import React from 'react'
import { render } from '@testing-library/react-native'
import { ThemeProvider } from 'styled-components'
import theme from '../theme'

const themeRender = (component: any) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}

export * from '@testing-library/react-native'

export { themeRender }
