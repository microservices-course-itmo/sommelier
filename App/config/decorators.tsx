import { ThemeProvider } from 'styled-components'
import React from 'react'
import styled from 'styled-components/native'
import { theme } from '../theme'

const StyledBufferView = styled.View`
  flex: 1;
  margin-vertical: 40;
  margin-horizontal: 20;
`

export const ThemeDecorator = (storyFn: any) => (
  <ThemeProvider theme={theme}>{storyFn()}</ThemeProvider>
)

export const BufferView = (storyFn: any) => (
  <StyledBufferView>{storyFn()}</StyledBufferView>
)
