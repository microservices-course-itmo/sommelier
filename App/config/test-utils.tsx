import React, { ReactNode } from 'react'
import { render } from '@testing-library/react-native'
import { ThemeProvider } from 'styled-components/native'
import { theme } from '../theme'

const themeRender = (component: ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}

export * from '@testing-library/react-native'

export { themeRender }
