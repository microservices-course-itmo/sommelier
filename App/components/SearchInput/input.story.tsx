import React, { useState } from 'react'
import styled from 'styled-components/native'
import { action } from '@storybook/addon-actions'
import { boolean, select, text, number } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react-native'
import { BufferView, ThemeDecorator } from '../../config/decorators'
import { SearchInput } from './input'
import { INPUT_SIZE_KEYS } from '../../theme/inputSearchSizes'

const StyledWrapper = styled.View`
  display: flex;
  flex-direction: column;
`

const StyledPadding = styled.View`
  margin-top: 5px;
  margin-bottom: 5px;
`

export const list = () => (
  <StyledWrapper>
    <SearchInput
      onChange={() => 's'}
      size='lg'
      clearButtonMode
      placeholder='Поиск по избранному'
      editable={false}
    />
    <StyledPadding />
    <SearchInput
      onChange={() => 's'}
      size='md'
      clearButtonMode
      placeholder='Поиск по избранному'
      editable
    />
    <StyledPadding />
    <SearchInput
      onChange={() => 's'}
      size='sm'
      clearButtonMode
      placeholder='Поиск по избранному'
      editable
    />
    <StyledPadding />
    <StyledPadding />
    <SearchInput
      onChange={() => 's'}
      size='lg'
      clearButtonMode={false}
      placeholder='Поиск по избранному'
      editable={false}
    />
    <StyledPadding />
    <SearchInput
      onChange={() => 's'}
      size='md'
      clearButtonMode={false}
      placeholder='Поиск по избранному'
      editable
    />
    <StyledPadding />
    <SearchInput
      onChange={() => 's'}
      size='sm'
      clearButtonMode={false}
      placeholder='Поиск по избранному'
      editable
    />
  </StyledWrapper>
)

export const InputKnobs = () => (
  <StyledWrapper>
    <SearchInput
      onChange={() => 'onChange called'}
      size={select('size', INPUT_SIZE_KEYS, INPUT_SIZE_KEYS.lg)}
      clearButtonMode={boolean('ClearButtonMode', false)}
      placeholder={text('Placeholder text', 'Поиск по избранному')}
      editable={boolean('Editable', true)}
      defaultValue={text('DefaultValue text', '')}
      value=''
      // maxLength={number('Max length', 15)}
    />
  </StyledWrapper>
)

storiesOf('Input', module)
  .addDecorator(BufferView)
  .addDecorator(ThemeDecorator)
  .add('Input list', list)

storiesOf('Input', module)
  .addDecorator(BufferView)
  .addDecorator(ThemeDecorator)
  .add('Input knobs', () => <InputKnobs />)
