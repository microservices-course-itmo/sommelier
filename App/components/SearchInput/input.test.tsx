import React from 'react'
import { themeRender, fireEvent } from '../../config/test-utils'
import SearchInput from './input'
import theme from '../../theme'

describe('SearchInput component', () => {
  const onChangeMock = jest.fn()
  it('Component with only necessary props matches snapshot', () => {
    const component = themeRender(<SearchInput onChange={onChangeMock} />)
    expect(component.toJSON()).toMatchSnapshot()
  })
  it('Component all props matches snapshot', () => {
    const component = themeRender(
      <SearchInput
        onChange={onChangeMock}
        defaultValue='abc'
        placeholder='123'
        editable
        size='lg'
        maxLength={30}
        value='987'
        clearButtonMode
      />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
  it('Default value prop assigns correctly', () => {
    const component = themeRender(
      <SearchInput onChange={onChangeMock} defaultValue='abc' />
    )
    expect(
      component.getByTestId('sommelier-search-input').props.defaultValue
    ).toBe('abc')
  })
  it('Placeholder prop assigns correctly', () => {
    const component = themeRender(
      <SearchInput onChange={onChangeMock} placeholder='abc' />
    )
    expect(
      component.getByTestId('sommelier-search-input').props.placeholder
    ).toBe('abc')
  })
  it('Editable prop assigns correctly', () => {
    const component = themeRender(
      <SearchInput onChange={onChangeMock} editable={false} />
    )
    expect(
      component.getByTestId('sommelier-search-input').props.editable
    ).toBeFalsy()
  })
  it('Size prop assigns correctly', () => {
    const component = themeRender(
      <SearchInput onChange={onChangeMock} size='md' clearButtonMode />
    )
    const searchSVG = component.getByTestId('sommelier-search-input-search-svg')
    const searchInput = component.getByTestId('sommelier-search-input')
    const closeSVG = component.getByTestId('sommelier-search-input-close-svg')
    expect(searchSVG.props.height).toBe(theme.searchSizes.md)
    expect(searchSVG.props.width).toBe(theme.searchSizes.md)
    expect(searchInput.props.size).toBe('md')
    expect(closeSVG.props.height).toBe(theme.inputSearchFontSizes.md)
    expect(closeSVG.props.height).toBe(theme.inputSearchFontSizes.md)
  })
  it('Value prop assigns correctly', () => {
    const component = themeRender(
      <SearchInput onChange={onChangeMock} value='123' />
    )
    expect(component.getByTestId('sommelier-search-input').props.value).toBe(
      '123'
    )
  })
  // TODO: OnChange testing
  // it('OnChange function should be called', () => {
  //   const component = themeRender(<SearchInput onChange={onChangeMock} />)
  //   const searchInput = component.getByTestId('sommelier-search-input')
  //   fireEvent.changeText(searchInput, '123')
  //   expect(onChangeMock).toHaveBeenCalled()
  // })
  it('Max length prop assigns correctly', () => {
    const component = themeRender(
      <SearchInput onChange={onChangeMock} maxLength={15} />
    )
    expect(
      component.getByTestId('sommelier-search-input').props.maxLength
    ).toBe(15)
  })
  it('Clear button mode works correctly', () => {
    const elem = themeRender(
      <SearchInput onChange={onChangeMock} clearButtonMode />
    )
    fireEvent.press(elem.getByTestId('sommelier-search-input-clear-button'))
    expect(elem.getByTestId('sommelier-search-input').props.value).toBe('')
  })
})
