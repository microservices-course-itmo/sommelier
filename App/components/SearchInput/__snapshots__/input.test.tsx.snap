// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component Component all props matches snapshot 1`] = `
<View
  editable={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderColor": "#9e9e9e",
        "borderRadius": 50,
        "borderStyle": "solid",
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "row",
        "overflow": "hidden",
        "paddingBottom": 2,
        "paddingLeft": 15,
        "paddingRight": 5,
        "paddingTop": 2,
        "width": "100%",
      },
    ]
  }
>
  <SvgMock
    height="30px"
    testID="sommelier-search-input-search-svg"
    width="30px"
  />
  <TextInput
    allowFontScaling={true}
    defaultValue="abc"
    editable={true}
    maxLength={30}
    onChange={[MockFunction]}
    onChangeText={[Function]}
    placeholder="123"
    rejectResponderTermination={true}
    size="lg"
    style={
      Array [
        Object {
          "fontSize": 18,
          "height": 50,
          "paddingLeft": 5,
          "paddingRight": 60,
          "width": "100%",
        },
      ]
    }
    testID="sommelier-search-input"
    underlineColorAndroid="transparent"
    value="987"
  />
  <View
    style={
      Array [
        Object {
          "height": "100%",
          "justifyContent": "center",
          "marginLeft": 5,
          "position": "absolute",
          "right": 15,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="sommelier-search-input-clear-button"
    >
      <SvgMock
        height="18px"
        testID="sommelier-search-input-close-svg"
        width="18px"
      />
    </View>
  </View>
</View>
`;

exports[`SearchInput component Component with only necessary props matches snapshot 1`] = `
<View
  editable={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderColor": "#9e9e9e",
        "borderRadius": 50,
        "borderStyle": "solid",
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "row",
        "overflow": "hidden",
        "paddingBottom": 2,
        "paddingLeft": 15,
        "paddingRight": 5,
        "paddingTop": 2,
        "width": "100%",
      },
    ]
  }
>
  <SvgMock
    height="20px"
    testID="sommelier-search-input-search-svg"
    width="20px"
  />
  <TextInput
    allowFontScaling={true}
    editable={true}
    onChange={[MockFunction]}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    size="md"
    style={
      Array [
        Object {
          "fontSize": 16,
          "height": 38,
          "paddingLeft": 5,
          "paddingRight": 60,
          "width": "100%",
        },
      ]
    }
    testID="sommelier-search-input"
    underlineColorAndroid="transparent"
  />
</View>
`;
